_posts/2017-07-26-introducing-workspaces.md

```diff
@@ -0,0 +1,237 @@
+    "jest-matcher-utils": "^20.0.3",
+    "pretty-format": "^20.0.3"
+  }
+}`
+```
+
+An approach that a wrapper like Lerna takes is to run `yarn install` for each package.json separately and then run `yarn link` for packages that depend on each other.
+
+We would get a folder structure like this.
+
+```
+`| jest/
+| ---- node_modules/
+| -------- chalk/
+| ---- package.json
+| ---- packages/
+| -------- jest-matcher-utils/
+| ------------ node_modules/
+`| ---------------- chalk/`
+`| ---------------- pretty-format/
+`| ------------ package.json
+| -------- jest-diff/
+| ------------ node_modules/
+``| ---------------- chalk/``
+`````| ---------------- diff/
+``````````| ---------------- jest-matcher-utils/  (symlink) -> ../jest-matcher-utils
+```| ---------------- pretty-format/
+``| ------------ package.json
+...`
+```
+
+As you see there is a redundancy of third-party dependencies.
+
+However, with Yarn Workspaces all node_modules are considered as one and Yarn tries to move all the packages to the root node_modules if there are no conflicts.
+
+```
+`| jest/
+| ---- node_modules/
+| -------- chalk/
+| -------- diff/
+| -------- pretty-format/
+`| -------- jest-matcher-utils/  (symlink) -> ../packages/jest-matcher-utils
+`| ---- package.json
+| ---- packages/
+| -------- jest-matcher-utils/
+| ------------ node_modules/
+`| ---------------- chalk/`
+| ------------ package.json
+| -------- jest-diff/
+| ------------ node_modules/
+``| ---------------- chalk/``
+| ------------ package.json
+...`
+```
+
+Packages `diff`, `pretty-format` and symlink to `jest-matcher-utils` were hoisted to the **root** node_modules making the installation faster and smaller.
+Package `chalk` could not be moved to the **root** node_modules because the **root** already depends on an different incompatible version of `chalk`.
+
+Both of the structures above are compatible but the latter one is more optimal while still being correct from the point of view of Node.js dependency resolution.
+
+If you run code inside `jest-diff` **workspace** it will be able to resolve all its dependencies: 
+
+* require('chalk') resolves to `./node_modules/chalk`
+* require('diff') resolves to `../../node_modules/diff`
+* require('pretty-format') resolves to `../../node_modules/pretty-format`
+* require('jest-matcher-utils') resolves to `../../node_modules/jest-matcher-utils`  that is a symlink to `../packages/jest-matcher-utils`
+
+## Managing dependencies of Workspaces
+
+If you want to add/remove/upgrade a dependency of a **workspace** just change run the appropriate command inside the **workspace** folder.
+
+```
+$ cd packages/jest-matcher-utils/
+$ yarn add left-pad
+✨ Done in 1.77s.
+$ git status
+modified: package.json
+modified: ../../yarn.lock
+```
+
+Note that **workspaces** don't have yarn.lock and the **root** yarn.lock contains all the dependencies for all the **workspaces**.
+
+## Integrating with Lerna
+
+Do Yarn Workspaces make Lerna obsolete?
+
+Not at all.
```

Comment:
This might be a bit misleading. While the current implementation of Yarn Workspaces might not totally replace the functionality of Lerna, the intention is that it will.

Yarn should solve many of the same problems and people should start adapting their tools to work on top of Yarn Workspaces instead of Lerna. I want to eventually deprecate Lerna altogether.